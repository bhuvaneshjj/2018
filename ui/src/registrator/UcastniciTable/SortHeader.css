.SortHeader_a {
  color: #4183c4;
  text-decoration: none;
  cursor: pointer;
}
