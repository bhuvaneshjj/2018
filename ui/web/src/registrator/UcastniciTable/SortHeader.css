.SortHeader {
  background-color: transparent;
  border: none;
  color: #4183c4;
  cursor: pointer;
}
