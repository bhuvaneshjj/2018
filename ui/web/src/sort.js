import moment from 'moment';

export const SortDirTypes = { NONE: 'none', ASC: 'asc', DESC: 'desc' };

export const reverseSortDirType = (sortDirType) => {
  switch (sortDirType) {
    case SortDirTypes.ASC:
      return SortDirTypes.DESC;
    case SortDirTypes.DESC:
      return SortDirTypes.ASC;
    default:
      return SortDirTypes.ASC;
  }
};

// Forces 'null' to be the last (desc = false) or first (desc = true).
// Note that 'undefined' is never passed to the sort method (virtue of Array.prototype.sort).
export const numberAndUndefinedSortMethod = (a, b, desc = false) => {
  if (a && b) {
    if (a < b) {
      return -1;
    }
    if (a > b) {
      return +1;
    }
    return 0; // a tie
  }

  if (!a && !b) {
    return 0; // both null or undefined
  }
  if (a) {
    return desc ? +1 : -1;
  }
  return desc ? -1 : +1;
};

export const narozeniSortMethod = (a, b, desc = false) => {
  if (a && b) {
    if (a.rok < b.rok) {
      return -1;
    }
    if (a.rok > b.rok) {
      return +1;
    }

    if (a.mesic && b.mesic) {
      if (a.mesic < b.mesic) {
        return -1;
      }
      if (a.mesic > b.mesic) {
        return +1;
      }

      return numberAndUndefinedSortMethod(a.den, b.den, desc);
    }

    if (!a.mesic && !b.mesic) {
      return 0; // both null or undefined
    }
    if (a.mesic) {
      return desc ? +1 : -1;
    }
    return desc ? -1 : +1;
  }

  if (!a && !b) {
    return 0; // both null
  }
  if (a) {
    return desc ? +1 : -1;
  }
  return desc ? -1 : +1;
};

const collator = new Intl.Collator('cs');
export const csStringSortMethod = (a, b) => collator.compare(a, b);

export const datumSortMethod = (a, b) => a.getTime() - b.getTime();

export const kategorieSortMethod = (a, b) => {
  let rv = csStringSortMethod(a.typ, b.typ);
  if (rv !== 0) {
    return rv;
  }

  if (a.pohlavi) {
    rv = csStringSortMethod(a.pohlavi, b.pohlavi);
    if (rv !== 0) {
      return rv;
    }
  }

  if (a.vek && b.vek) {
    return a.vek.min - b.vek.min;
  }

  return 0;
};

export const prijmeniJmenoNarozeniSortMethod = (a, b, desc = false) => {
  const prijmeniCmp = csStringSortMethod(a.prijmeni, b.prijmeni);
  if (prijmeniCmp !== 0) {
    return prijmeniCmp;
  }

  const jmenoCmp = csStringSortMethod(a.jmeno, b.jmeno);
  if (jmenoCmp !== 0) {
    return jmenoCmp;
  }

  return narozeniSortMethod(a.narozeni, b.narozeni, desc);
};

export const dokoncenoSortMethod = (a, b) => {
  if (a === b) {
    return 0;
  }

  if (a === undefined || a === null) {
    return +1;
  }
  if (b === undefined || b === null) {
    return -1;
  }

  if (a === true && b === false) {
    return -1;
  }
  return +1; // a === false, b === true
};

export const casSortMethod = (a, b, desc = false) => {
  const c = a ? moment.duration(a).asMilliseconds() : undefined;
  const d = b ? moment.duration(b).asMilliseconds() : undefined;
  return numberAndUndefinedSortMethod(c, d, desc);
};

export const dokoncenoCasSortMethod = (a, b, desc = false) => {
  const dokoncenoCmp = dokoncenoSortMethod(a.dokonceno, b.dokonceno);
  if (dokoncenoCmp !== 0) {
    return dokoncenoCmp;
  }

  return casSortMethod(a.cas, b.cas, desc);
};

export const sortForColumn = ({ data, sortColumn, sortDir, extraSortMethods = {} }) => {
  const desc = sortDir === SortDirTypes.DESC;

  const sortMethods = {
    absPoradi: (a, b) => numberAndUndefinedSortMethod(a.absPoradi, b.absPoradi, desc),
    cas: (a, b) => casSortMethod(a.cas, b.cas, desc),
    datum: (a, b) => datumSortMethod(a.datum, b.datum),
    dokonceno: (a, b) => dokoncenoSortMethod(a.dokonceno, b.dokonceno),
    email: (a, b) => csStringSortMethod(a.email, b.email),
    jmeno: (a, b) => csStringSortMethod(a.jmeno, b.jmeno),
    kategorie: (a, b) => kategorieSortMethod(a.kategorie, b.kategorie),
    narozeni: (a, b) => narozeniSortMethod(a.narozeni, b.narozeni, desc),
    obec: (a, b) => csStringSortMethod(a.obec, b.obec),
    prijmeni: (a, b) => csStringSortMethod(a.prijmeni, b.prijmeni),
    relPoradi: (a, b) => numberAndUndefinedSortMethod(a.relPoradi, b.relPoradi, desc),
    startCislo: (a, b) => numberAndUndefinedSortMethod(a.startCislo, b.startCislo, desc),
    zaplaceno: (a, b) => a.zaplaceno - b.zaplaceno,
    ...extraSortMethods,
  };

  const sortMethod = sortMethods[sortColumn] || prijmeniJmenoNarozeniSortMethod;
  const sorted = data.slice().sort(sortMethod);

  if (desc) {
    sorted.reverse();
  }

  return sorted;
};
