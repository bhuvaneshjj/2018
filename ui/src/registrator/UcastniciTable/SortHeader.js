import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { SortDirTypes } from '../../sort';
import './SortHeader.css';

const sortDirs = {
  [SortDirTypes.NONE]: '',
  [SortDirTypes.ASC]: '↑',
  [SortDirTypes.DESC]: '↓'
};

class SortHeader extends PureComponent {
  handleClick = event => {
    event.preventDefault();
    this.props.onClick();
  };

  render = () => (
    <a className="SortHeader_a" onClick={this.handleClick}>
      {this.props.children} {sortDirs[this.props.sortDir]}
    </a>
  );
}

SortHeader.propTypes = {
  children: PropTypes.node,
  sortDir: PropTypes.oneOf([SortDirTypes.NONE, SortDirTypes.ASC, SortDirTypes.DESC]),
  onClick: PropTypes.func.isRequired
};

export default SortHeader;
