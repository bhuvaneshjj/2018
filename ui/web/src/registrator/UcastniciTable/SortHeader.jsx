import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { SortDirTypes } from '../../sort';
import './SortHeader.css';

const sortDirs = {
  [SortDirTypes.NONE]: '',
  [SortDirTypes.ASC]: '↑',
  [SortDirTypes.DESC]: '↓'
};

class SortHeader extends PureComponent {
  handleClick = event => {
    event.preventDefault();
    this.props.onClick();
  };

  render = () => (
    <button className="SortHeader" onClick={this.handleClick}>
      {this.props.children} {sortDirs[this.props.sortDir]}
    </button>
  );
}

SortHeader.propTypes = {
  children: PropTypes.node,
  sortDir: PropTypes.oneOf([SortDirTypes.NONE, SortDirTypes.ASC, SortDirTypes.DESC]),
  onClick: PropTypes.func.isRequired
};

SortHeader.defaultProps = {
  children: undefined,
  sortDir: SortDirTypes.NONE
};

export default SortHeader;
