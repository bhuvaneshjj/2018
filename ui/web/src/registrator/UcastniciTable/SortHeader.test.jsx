import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import { SortDirTypes } from '../../sort';
import SortHeader from './SortHeader';

it('renders the default', () => {
  const component = renderer.create(
    <SortHeader onClick={jest.fn()} sortDir={SortDirTypes.NONE}>
      příjmení
    </SortHeader>
  );
  expect(component.toJSON()).toMatchSnapshot();
});

it('renders jméno sestupně', () => {
  const component = renderer.create(
    <SortHeader onClick={jest.fn()} sortDir={SortDirTypes.DESC}>
      jméno
    </SortHeader>
  );
  expect(component.toJSON()).toMatchSnapshot();
});

it('maps onClick to dispatch onClick action', () => {
  const onClick = jest.fn();
  const wrapper = mount(
    <SortHeader onClick={onClick} sortDir={SortDirTypes.ASC}>
      jméno
    </SortHeader>
  );
  expect(wrapper.find('button')).toHaveLength(1);
  wrapper.find('button').simulate('click');

  expect(onClick).toHaveBeenCalledTimes(1);
});
