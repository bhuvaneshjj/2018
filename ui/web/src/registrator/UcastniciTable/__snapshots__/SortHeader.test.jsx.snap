// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders jméno sestupně 1`] = `
<button
  className="SortHeader"
  onClick={[MockFunction]}
  type="button"
>
  jméno
   
  ↓
</button>
`;

exports[`renders the default 1`] = `
<button
  className="SortHeader"
  onClick={[MockFunction]}
  type="button"
>
  příjmení
   
  
</button>
`;
